geomstats/backend/pytorch.py

```diff
@@ -0,0 +1,318 @@
+"""Pytorch based computation backend."""
+
+import numpy as np
+import torch
+
+int8 = 'torch.ByteTensor'
+int32 = 'torch.LongTensor'
+double = 'torch.DoubleTensor'
+float64 = 'torch.DoubleTensor'
+float16 = 'torch.Float'
+
+
+def arctan2(*args, **kwargs):
+    return torch.arctan2(*args, **kwargs)
+
+
+def cast(x, dtype):
+    return x.type(dtype)
+
+
+def divide(*args, **kwargs):
+    return torch.div(*args, **kwargs)
+
+
+def repeat(*args, **kwargs):
+    return torch.repeat(*args, **kwargs)
+
+
+def asarray(x):
+    return np.asarray(x)
+
+
+def concatenate(*args, **kwargs):
+    return torch.concatenate(*args, **kwargs)
+
+
+def identity(val):
+    return torch.identity(val)
+
+
+def hstack(val):
+    return torch.hstack(val)
+
+
+def stack(*args, **kwargs):
+    return torch.stack(*args, **kwargs)
+
+
+def vstack(val):
+    return torch.vstack(val)
+
+
+def array(val):
+    return torch.tensor(val)
+
+
+def abs(val):
+    return torch.abs(val)
+
+
+def zeros(*args):
+    return torch.from_numpy(np.zeros(*args)).float()
+
+
+def ones(val):
+    return torch.ones(val)
+
+
+def ones_like(*args, **kwargs):
+    return torch.ones_like(*args, **kwargs)
+
+
+def empty_like(*args, **kwargs):
+    return torch.empty_like(*args, **kwargs)
+
+
+def all(x, axis=None):
+    if axis is None:
+        return x.byte().all()
+    return torch.from_numpy(np.all(x, axis=axis).astype(int))
+
+
+def allclose(a, b, **kwargs):
+    return torch.allclose(a, b, **kwargs)
+
+
+def sin(val):
+    return torch.sin(val)
+
+
+def cos(val):
+    return torch.cos(val)
+
+
+def cosh(*args, **kwargs):
+    return torch.cosh(*args, **kwargs)
+
+
+def sinh(*args, **kwargs):
+    return torch.sinh(*args, **kwargs)
+
+
+def tanh(*args, **kwargs):
+    return torch.tanh(*args, **kwargs)
+
+
+def arcsinh(x):
+    return torch.log(x + torch.sqrt(x*x+1))
+
+
+def arcosh(x):
+    return torch.log(x + torch.sqrt(x*x-1))
+
+
+def tan(val):
+    return torch.tan(val)
+
+
+def arcsin(val):
+    return torch.asin(val)
+
+
+def arccos(val):
+    return torch.acos(val)
+
+
+def shape(val):
+    return val.shape
+
+
+def dot(a, b):
+    return torch.from_numpy(np.dot(a, b))
+
+
+def maximum(a, b):
+    return torch.max(array(a), array(b))
```

Comment:
why array(a) and array(b)?

Reply:
if not cast to an array it will fail trying to compare values